qualtran/bloqs/hamiltonian_simulation/hamiltonian_simulation_by_gqsp.py

```diff
@@ -0,0 +1,197 @@
+
+    References:
+        [Generalized Quantum Signal Processing](https://arxiv.org/abs/2308.01501)
+            Motlagh and Wiebe. (2023). Theorem 7, Corollary 8.
+
+    Args:
+        walk_operator: qubitization walk operator of $H$ constructed from SELECT and PREPARE oracles.
+        t: time to simulate the Hamiltonian, i.e. $e^{-iHt}$
+        alpha: the $1$-norm of the coefficients of the unitaries comprising the Hamiltonian $H$.
+        precision: the precision $\epsilon$ to approximate $e^{it\cos\theta}$ to a polynomial.
+    """
+
+    walk_operator: QubitizationWalkOperator
+    t: SymbolicFloat = field(kw_only=True)
+    alpha: SymbolicFloat = field(kw_only=True)
+    precision: SymbolicFloat = field(kw_only=True)
+
+    def _parameterized_(self):
+        return is_symbolic(self.t, self.alpha, self.precision)
+
+    @cached_property
+    def degree(self) -> SymbolicInt:
+        r"""degree of the polynomial to approximate the function e^{it\cos(\theta)}"""
+        return degree_jacobi_anger_approximation(self.t * self.alpha, precision=self.precision)
+
+    @cached_property
+    def approx_cos(self) -> NDArray[np.complex_]:
+        r"""polynomial approximation for $$e^{i\theta} \mapsto e^{it\cos(\theta)}$$"""
+        if self._parameterized_():
+            raise ValueError(f"cannot compute `cos` approximation for parameterized Bloq {self}")
+        poly = approx_exp_cos_by_jacobi_anger(-self.t * self.alpha, degree=self.degree)
+        poly = scale_down_to_qsp_polynomial(poly) * (1 - 2 * self.precision)
+        return poly
+
+    @cached_property
+    def gqsp(self) -> GeneralizedQSP:
+        return GeneralizedQSP.from_qsp_polynomial(
+            self.walk_operator,
+            self.approx_cos,
+            negative_power=self.degree,
+            verify=True,
+            verify_precision=1e-4,
+        )
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return self.gqsp.signature
+
+    def __add_prepare(
+        self,
+        bb: 'BloqBuilder',
+        gqsp_soqs: Dict[str, 'SoquetT'],
+        state_prep_ancilla_soqs: Dict[str, 'SoquetT'],
+        *,
+        adjoint: bool = False,
+    ) -> Tuple[Dict[str, 'SoquetT'], Dict[str, 'SoquetT']]:
+        prepare = self.walk_operator.prepare
+
+        selection_registers = {reg.name: gqsp_soqs[reg.name] for reg in prepare.selection_registers}
+        prepare_out_soqs = bb.add_d(
+            prepare.adjoint() if adjoint else prepare,
+            **selection_registers,
+            **state_prep_ancilla_soqs,
+        )
+        gqsp_soqs |= {
+            reg.name: prepare_out_soqs.pop(reg.name) for reg in prepare.selection_registers
+        }
+        return gqsp_soqs, prepare_out_soqs
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> Dict[str, 'SoquetT']:
+        state_prep_ancilla = {
+            reg.name: bb.allocate(reg.total_bits())
+            for reg in self.walk_operator.prepare.junk_registers
+        }
```

Comment:
I think this wouldn't work with the cirq style API but we can fix it later.

`Allocate` and `Free` bloqs need to implement `as_cirq_op`. The reason it works right now is probably because there are no junk registers in the tested bloq so no new qubits are being allocated as part of the decomposition. Let's open an issue with a minimal example to reproduce.

cc @skushnir123 has run into this as well.

Reply:
for the example, which cirq function should I test against? Because `cirq.unitary` doesn't make sense if there are junk registers (or does it?).

Reply:
Using `allocate` and `free` does seem to work with cirq?
```py
from qualtran import BloqBuilder
from qualtran.bloqs.basic_gates import Hadamard

bb = BloqBuilder()
q = bb.allocate(1)
q = bb.add(Hadamard(), q=q)
bb.free(q)
bloq = bb.finalize()

bloq.to_cirq_circuit()
```

```
(_c(0): ───alloc───H───free───, {})
```

Reply:
you can test against generating the cirq circuit for the decomposition. So `cirq.Circuit(cirq.decompose_once(GateHelper(bloq).operation))`

Reply:
I made a simple bloq which allocates and frees a qubit. The cirq conversion seems to work as expected.

```py
import cirq
from qualtran import BloqBuilder, GateWithRegisters, Signature
from qualtran.bloqs.basic_gates import CNOT, TGate
from qualtran.cirq_interop.testing import GateHelper


class MyBloq(GateWithRegisters):
    @property
    def signature(self) -> Signature:
        return Signature.build(q=1)

    def build_composite_bloq(self, bb: 'BloqBuilder', q) -> dict[str, 'SoquetT']:
        a = bb.allocate(1)
        q, a = bb.add(CNOT(), ctrl=q, target=a)
        q = bb.add(TGate(), q=q)
        bb.free(a)
        return {'q': q}


print(cirq.Circuit(cirq.decompose_once(GateHelper(MyBloq()).operation)))
```

ouputs:
```
_decompose_protocol_0_c(0): ───alloc───X───free───
                                       │
q: ────────────────────────────────────@───T──────
```